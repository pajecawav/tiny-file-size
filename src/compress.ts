import zlib from "zlib";

export function getGzipSize(buffer: Buffer): Promise<number> {
	return new Promise((resolve, reject) => {
		zlib.gzip(buffer, (error, result) => {
			if (error) return reject(error);
			resolve(result.length);
		});
	});
}

export function getBrotliSize(buffer: Buffer): Promise<number> {
	return new Promise((resolve, reject) => {
		zlib.brotliCompress(buffer, (error, result) => {
			if (error) return reject(error);
			resolve(result.length);
		});
	});
}
