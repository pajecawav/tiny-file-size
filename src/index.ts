export { getBrotliSize, getGzipSize, getRawSize } from "./compress";
export { buildJsonReport, buildPrettyReport, humanizeBytes } from "./report";
export type { PrettyReportOptions } from "./report";
export { getFileSizes, run } from "./run";
export type { FileSize, FileSizesOptions } from "./run";
