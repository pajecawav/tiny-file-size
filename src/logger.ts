import colors from "picocolors";

export const logger = {
	error(message: string) {
		console.error(colors.bgRed(colors.bold(colors.black(" ERROR "))), colors.red(message));
	},
};
