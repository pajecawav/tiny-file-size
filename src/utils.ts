import binaryExtensions from "binary-extensions";
import { extname } from "path";

const binaryExtensionsSet = new Set(binaryExtensions);

export function sum(values: number[]): number {
	let total = 0;
	for (const value of values) total += value;
	return total;
}

export function isBinaryFile(path: string): boolean {
	return binaryExtensionsSet.has(extname(path).slice(1));
}
