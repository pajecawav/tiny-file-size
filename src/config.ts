import colors from "picocolors";
import { version } from "../package.json";
import { logger } from "./logger";

export interface Config {
	globs: string[];
	gzip: boolean;
	brotli: boolean;
	output: "json" | "pretty";
	total: boolean;
}

const HELP = `
${colors.bold("Usage:")}
  file-size [options] [globs...]

${colors.bold("Options:")}
  -h, --help                     Show help
  -g, --gzip                     Include gzip size in output 
  -b, --brotli                   Include brotli size in output 
      --total                    Include total size in output
      --json                     Print result as json
  -v, --version                  Print the current version and exit

${colors.bold("Examples:")}
  # Show sizes of files foo.js and bar.json
  file-size foo.js bar.json

  # Include sizes after gzip and brotli
  file-size --gzip --brotli foo.js bar.json

  # Print output as JSON
  file-size --json foo.js bar.json
`.trim();

export function parseArgs(args: string[]): Config {
	const cfg: Config = {
		globs: [],
		gzip: false,
		brotli: false,
		output: "pretty",
		total: false,
	};

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];

		if (arg[0] !== "-") {
			cfg.globs = args.slice(i);
			break;
		}

		switch (arg) {
			case "-h":
			case "--help":
				console.log(HELP);
				process.exit(0);
				break;
			case "-v":
			case "--version":
				console.log(version);
				process.exit(0);
				break;
			case "-g":
			case "--gzip":
				cfg.gzip = true;
				break;
			case "-b":
			case "--brotli":
				cfg.brotli = true;
				break;
			case "--json":
				cfg.output = "json";
				break;
			case "--total":
				cfg.total = true;
				break;
			default:
				logger.error(`Unknown argument ${arg}`);
				process.exit(1);
		}
	}

	return cfg;
}

export const config = parseArgs(process.argv.slice(2));
