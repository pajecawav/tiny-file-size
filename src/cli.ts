import { config } from "./config";
import { run } from "./run";

run(config);
