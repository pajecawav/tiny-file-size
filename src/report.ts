import path from "path";
import { bold, dim, green } from "picocolors";
import { FileSize } from "./run";

export function reportJson(sizes: FileSize[]) {
	console.log(sizes);
}

export function reportConsole(sizes: FileSize[]) {
	for (const { file, plain, gzip, brotli } of sizes) {
		const { dir, base } = path.parse(file);

		const pathStr = (dir ? dim(dir + "/") : "") + green(base);

		const plainStr = humanizeBytes(plain);
		const gzipStr = humanizeBytes(gzip);
		const brotliStr = humanizeBytes(brotli);

		console.log(`${pathStr}: ${bold(plainStr)} (${gzipStr} gzip, ${brotliStr} brotli)`);
	}
}

function humanizeBytes(bytes: number): string {
	let suffix = "MB";
	for (const suff of ["B", "KB"]) {
		if (bytes < 1024) {
			suffix = suff;
			break;
		}
		bytes /= 1024;
	}

	if (suffix === "B") {
		return `${bytes.toFixed(0)} ${suffix}`;
	}

	return `${bytes.toFixed(2)} ${suffix}`;
}
