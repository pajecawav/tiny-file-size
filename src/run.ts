import { existsSync, statSync } from "fs";
import { globbySync, isDynamicPattern } from "globby";
import { getBrotliSize, getGzipSize, getRawSize } from "./compress";
import { Config } from "./config";
import { logger } from "./logger";
import { buildJsonReport, buildPrettyReport } from "./report";

export interface FileSize {
	file: string;
	raw: number;
	gzip: number | null;
	brotli: number | null;
}

interface FileSizesOptions {
	files: string[];
	gzip: boolean;
	brotli: boolean;
}

export async function getFileSizes(config: FileSizesOptions): Promise<FileSize[]> {
	const sizes: FileSize[] = await Promise.all(
		config.files.map(async file => {
			const [raw, gzip, brotli] = await Promise.all([
				getRawSize(file),
				config.gzip ? getGzipSize(file) : null,
				config.brotli ? getBrotliSize(file) : null,
			]);

			return { file, raw, gzip, brotli };
		})
	);

	return sizes;
}

export function getFilesFromGlobs(globs: string[]): string[] {
	const filesSet = new Set<string>();

	for (const globOrFile of globs) {
		// each
		if (isDynamicPattern(globOrFile)) {
			for (const file of globbySync(globOrFile)) {
				filesSet.add(file);
			}
		} else {
			if (!existsSync(globOrFile) || !statSync(globOrFile).isFile()) {
				throw new Error(`No such file ${globOrFile}`);
			}

			filesSet.add(globOrFile);
		}
	}

	return Array.from(filesSet);
}

export async function run(config: Config) {
	let files: string[];
	try {
		files = getFilesFromGlobs(config.files);
	} catch (e: unknown) {
		if (!(e instanceof Error)) {
			throw e;
		}

		logger.error(e.message);
		process.exit(1);
	}

	let sizes: FileSize[];
	try {
		sizes = await getFileSizes({ ...config, files });
	} catch (e: unknown) {
		if (!(e instanceof Error)) {
			throw e;
		}

		logger.error(e.message);
		process.exit(1);
	}

	let report: string;
	switch (config.output) {
		case "json":
			report = buildJsonReport(sizes);
			break;
		case "pretty":
			report = buildPrettyReport(sizes, { total: config.total });
			break;
	}

	console.log(report);
}
