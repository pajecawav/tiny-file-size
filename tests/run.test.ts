import { describe, expect, it } from "vitest";
import { getFilesFromGlobs, getFileSizes } from "../src/run";
import { files, resolveDataFixture } from "./fixtures/files";

describe("getFileSizes", () => {
	it("counts raw sizes by default", async () => {
		const sizes = await getFileSizes({ files, gzip: false, brotli: false });
		expect(sizes).toMatchSnapshot();
	});

	it("counts gzip and brotli sizes", async () => {
		const sizes = await getFileSizes({ files, gzip: true, brotli: true });
		expect(sizes).toMatchSnapshot();
	});
});

describe("getFilesFromGlobs", () => {
	it("supports globs", async () => {
		const files = await getFilesFromGlobs([resolveDataFixture("*.{md,txt}")]);
		expect(files).toMatchSnapshot();
	});

	it("treats strings without dynamic patterns as paths", async () => {
		const files = await getFilesFromGlobs([
			resolveDataFixture("lorem.txt"),
			resolveDataFixture("*.md"),
		]);
		expect(files).toMatchSnapshot();
	});

	it("throws when file doesn't exist", async () => {
		const fn = () => {
			getFilesFromGlobs([resolveDataFixture("lorem.txt"), resolveDataFixture("doesntexist")]);
		};

		await expect(fn).toThrow(/no such file/i);
	});
});
