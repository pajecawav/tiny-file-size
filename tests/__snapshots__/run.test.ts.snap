// Vitest Snapshot v1

exports[`getFileSizes > counts gzip and brotli sizes 1`] = `
[
  {
    "brotli": 1395,
    "file": "tests/fixtures/data/foo.md",
    "gzip": 1484,
    "raw": 2519,
  },
  {
    "brotli": 2193,
    "file": "tests/fixtures/data/lorem.md",
    "gzip": 2410,
    "raw": 4183,
  },
  {
    "brotli": 1142,
    "file": "tests/fixtures/data/lorem.txt",
    "gzip": 1205,
    "raw": 2877,
  },
]
`;

exports[`getFileSizes > counts raw sizes by default 1`] = `
[
  {
    "brotli": null,
    "file": "tests/fixtures/data/foo.md",
    "gzip": null,
    "raw": 2519,
  },
  {
    "brotli": null,
    "file": "tests/fixtures/data/lorem.md",
    "gzip": null,
    "raw": 4183,
  },
  {
    "brotli": null,
    "file": "tests/fixtures/data/lorem.txt",
    "gzip": null,
    "raw": 2877,
  },
]
`;

exports[`getFilesFromGlobs > supports globs 1`] = `
[
  "tests/fixtures/data/foo.md",
  "tests/fixtures/data/lorem.md",
  "tests/fixtures/data/lorem.txt",
]
`;

exports[`getFilesFromGlobs > treats strings without dynamic patterns as paths 1`] = `
[
  "tests/fixtures/data/lorem.txt",
  "tests/fixtures/data/foo.md",
  "tests/fixtures/data/lorem.md",
]
`;
